File: app/src/lib/api/startup.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * A single entry in the app's known-layout registry (Spec 013 / S5).
 *
 * Mirrors the Rust `KnownLayoutEntry` shape (serialised camelCase).
 * The layout picker (S6) displays `name`, `path`, and `lastOpened`.
 */
export interface KnownLayoutEntry {
  /** Display name shown in the layout picker. */
  name: string;
  /** Absolute path to the `.layout` base file. */
  path: string;
  /** ISO 8601 timestamp of the most recent open. */
  lastOpened: string;
}

/**
 * Read the known-layout registry from `$APPDATA/bowties/known-layouts.json`.
 *
 * Entries whose `path` no longer exists on disk are filtered out
 * before being returned, so the picker never shows a layout the
 * user can't actually open.
 */
export async function getKnownLayouts(): Promise<KnownLayoutEntry[]> {
  return invoke<KnownLayoutEntry[]>('get_known_layouts');
}

/**
 * Add or refresh a known-layout entry. Entries are matched by
 * `path` — re-adding a path replaces the existing entry's name and
 * `lastOpened` timestamp without duplicating it.
 *
 * @returns The post-write registry (with stale entries filtered).
 */
export async function addKnownLayout(
  entry: KnownLayoutEntry,
): Promise<KnownLayoutEntry[]> {
  return invoke<KnownLayoutEntry[]>('add_known_layout', { entry });
}

/**
 * Remove a known-layout entry by path. The `.layout` file and its
 * companion directory on disk are not touched — only the registry
 * forgets the path.
 *
 * @returns The post-write registry (with stale entries filtered).
 */
export async function removeKnownLayout(
  path: string,
): Promise<KnownLayoutEntry[]> {
  return invoke<KnownLayoutEntry[]>('remove_known_layout', { path });
}

File: app/src-tauri/src/commands/startup.rs
//! Startup / known-layout registry commands (Spec 013 / S5).
//!
//! Thin Tauri wrappers over [`crate::layout::known_layouts`]. The
//! domain module owns the on-disk format, the stale-path filter,
//! and the atomic write protocol; these commands only resolve the
//! `$APPDATA/bowties/known-layouts.json` path and translate errors
//! to strings for the IPC boundary.

use std::path::PathBuf;
use tauri::{AppHandle, Manager};

use crate::layout::known_layouts::{
    self, KnownLayoutEntry,
};

/// Resolve `$APPDATA/bowties/known-layouts.json` (or the platform
/// equivalent). Errors here are surfaced verbatim so the picker can
/// fall back to an empty list without crashing.
fn registry_path(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to get app data dir: {}", e))?;
    Ok(dir.join("known-layouts.json"))
}

/// Return the known-layout registry, with stale entries filtered.
#[tauri::command]
pub fn get_known_layouts(app: AppHandle) -> Result<Vec<KnownLayoutEntry>, String> {
    let path = registry_path(&app)?;
    Ok(known_layouts::load_known_layouts(&path))
}

/// Add or refresh a known-layout entry (matched by `path`).
/// Returns the post-write filtered registry.
#[tauri::command]
pub fn add_known_layout(
    app: AppHandle,
    entry: KnownLayoutEntry,
) -> Result<Vec<KnownLayoutEntry>, String> {
    let registry = registry_path(&app)?;
    known_layouts::add_known_layout(&registry, entry)
}

/// Remove a known-layout entry by `path`. Does not touch the
/// `.layout` file or its companion directory.
#[tauri::command]
pub fn remove_known_layout(
    app: AppHandle,
    path: String,
) -> Result<Vec<KnownLayoutEntry>, String> {
    let registry = registry_path(&app)?;
    known_layouts::remove_known_layout(&registry, &path)
}
